### ISSUE TITLE
`--watch-config` should watch enrichment table files

### ISSUE THREAD

**Issue by Jacq** [vectordotdev/vector#11817]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Use Cases

Using the plugin for geoip https://vector.dev/docs/reference/configuration/transforms/geoip/ I did not found any comments related to the update of this database while vector is running, is the new data automatically reloaded?

I updated the maxmind database file while vector was running, 

### Attempted Solutions

I just performed manually a restart of the vector to ensure the new maxmind database is used by vector, to make this more automatic I schedule a crontask to download the database and if successfully it restart the vector process.

### Proposal

It will be great if vector reload this database information automatically as soon the files changes which can be enabled with a command line option similar to the current --watch for the config.toml
It will be even greater if you can supply the maxmind user id/license and cron schedule so vector performs the database download automatically.

**Comment by jszwedko** [comment]
Hi @Jacq !

Currently the database is only loaded when Vector starts. I agree that we should tie this into Vector's reload functionality for consistency.

**Comment by onedr0p** [comment]
I am chiming in to say a reload on vector would be great when the database file changes, I think having vector manage the database for us and update / download is a bit too much though. There's tools you can use outside of Vector to automate that task.

**Comment by onedr0p** [comment]
Would it also be in scope here to not have vector crash is the geoip database cannot be found and instead rely on eventual consistency? I have a k8s cronjob that runs and store the geoip database in a RWX volume, if vector starts and that volume doesn't have the geoip database it crashes hard.

**Comment by jszwedko** [comment]
> Would it also be in scope here to not have vector crash is the geoip database cannot be found and instead rely on eventual consistency? I have a k8s cronjob that runs and store the geoip database in a RWX volume, if vector starts and that volume doesn't have the geoip database it crashes hard.

Yeah, we could. What would you want to happen with events flowing through when the geoip database couldn't be loaded? Would you just want them to flow through without the geoip enrichment? Or would you want it to block until it successfully loaded the database?

**Comment by onedr0p** [comment]
Blocking might be a bad idea IMO, I think a warning in the vector logs would suffice.

**Comment by jszwedko** [comment]
> Blocking might be a bad idea IMO, I think a warning in the vector logs would suffice.

👍 Warning and just passing events through unannotated you mean?

**Comment by onedr0p** [comment]
Right, that would be stellar!

**Comment by sooslaca** [comment]
Hi @Jacq , nice proposal, I upvoted. 

By any chance instead of restarting vector did you try send SIGHUP to the process and see if geoip db reloads too? 

Thanks

**Comment by sooslaca** [comment]
I tested by myself, you can reload it with SIGHUP when mmdb file time is newer that the old one.

**Comment by D13410N3** [comment]
Hello! Is there any ETA of this feature?

**Comment by jszwedko** [comment]
Hi @D13410N3 !

Unfortunately not yet. The reload can be triggered via SIGHUP as a workaround though.

Happy to help shepherd a PR through for this if anyone wants to take a shot 🙂

**Comment by valleedelisle** [comment]
+1 on this issue. That would be awesome to see this implemented.

**Comment by j4mbob** [comment]
+1, would be a very welcome change.

**Comment by Rentu** [comment]
+1

**Comment by mohan-dhayalan** [comment]
It will be really useful for us to have this feature implemented. Is there a workaround to reload CSV without restarting the vector?

**Comment by pront** [comment]
> It will be really useful for us to have this feature implemented. Is there a workaround to reload CSV without restarting the vector?

I am not aware of any workaround on this. However, there was some activity in area [here](https://github.com/vectordotdev/vector/issues/17283). Support for enrichment table files will require similar work.

**Comment by jszwedko** [comment]
> It will be really useful for us to have this feature implemented. Is there a workaround to reload CSV without restarting the vector?

Note that you can reload the CSV by issuing a SIGHUP to Vector to trigger a reload. This issue is about watching for changes automatically.

**Comment by esensar** [comment]
I wanted to pick this one up, but I am not sure what approach to take here.

Currently, all tables except for `memory` are read-only - they are created when configuration is changed (or Vector is just started).

We could make them all more similar to `memory` and have their implementation be backed by something like [evmap](https://docs.rs/evmap/latest/evmap/), which is used for `memory`, but it should support lookups by more than one key.

Maybe an easier option would be to somehow check more often if tables need a reload (or to watch filesystem) and then handle it via `TableRegistry`. It seems to me like it can handle loading new tables while old tables are still used pretty well.

**Comment by pront** [comment]
Hi @esensar, I am very happy to hear you are interested in this enhancement. 

From a design perspective, I would expect something similar with what we do for config watching:
 
* a component keeps the relevant table state and can accept signals to reload a certain table
* a watcher who can maintain a diff set and can send the signal

This might be a bit too high level but it leaves choices up to the implementer and it's not restrictive.

P.S. Things like `evmap` sounds like a step in the wrong direction for this kind of work. But maybe you had a more complex use case in mind.

**Comment by esensar** [comment]
> P.S. Things like `evmap` sounds like a step in the wrong direction for this kind of work. But maybe you had a more complex use case in mind.

Yes, agreed. It would just introduce additional complexity, with no real value (considering that `TableRegistry` can handle reloads without blocking anyways).
